kiwi_lint/similar_string.py

```diff
@@ -0,0 +1,146 @@
+
+            for file_name in files:
+                if file_name.endswith(".html"):
+                    self.all_html_files.add(
+                        os.path.join(self.project_root, rootpath, file_name)
+                    )
+
+    @staticmethod
+    def clean_string(text):
+        """
+        This method removes the operators and other punctuations
+        used in the string to avoid unwanted data to add up to
+        the similarity between the strings.
+        """
+        cleaned = ""
+        for char in text:
+            if char not in string.punctuation:
+                cleaned += char
+        return cleaned
+
+    def check_similar_string(self, translation_string):
+        cleaned_translation_string = self.clean_string(translation_string)
+        for key in self._dict_of_strings:
+            if (
+                levenshtein.normalized_similarity(
+                    cleaned_translation_string, self.clean_string(key)
+                )
+                >= self.threshold
+            ):
+                return key
+        return None
+
+    @utils.check_messages("similar-string")
+    def visit_call(self, node):
+        if not (
+            isinstance(node.func, astroid.Name)
+            and node.func.name in ("_", "gettext_lazy")
+        ):
+            return
+        if not isinstance(node.args[0], astroid.nodes.Const):
+            return
+
+        translation_string = node.args[0].value
+
+        if translation_string in self._dict_of_strings:
+            return
+
+        similar_string = self.check_similar_string(translation_string)
+
+        if similar_string:
+            self.add_message(
+                "similar-string",
+                node=node,
+                args=(
+                    translation_string,
+                    similar_string,
+                ),
+            )
+            return
+
+        self._dict_of_strings[translation_string] = True
+
+    def parse_translation_string(self, filename, lines):
+        for lineno, line in enumerate(lines):
+            match_in_line = re.finditer(r'{% trans "([\w ]+)" %}', line)
+            for match in match_in_line:
+                if filename not in self.error_messages:
+                    self.error_messages[filename] = []
+
+                translation_string = match.group(1)
+
+                if translation_string in self._dict_of_strings:
+                    continue
+
+                similar_string = self.check_similar_string(translation_string)
+
+                if similar_string:
+                    self.error_messages[filename].append(
+                        (
+                            lineno + 1,
+                            match.start() + 10,
+                            translation_string,
+                            similar_string,
+                        )
+                    )
+                    continue
+
+                self._dict_of_strings[translation_string] = True
```

Comment:
I don't understand the if-continue from the previous block + this line. Can you walk me through the code here.

Reply:
It is very similar to the if-return from the `visit_call` method, but here we're parsing each string we found in loop instead of function as in previous, so, the first if-continue states that if the same string is found in `_dict_of_strings`, then we should not go to next line instead continue to next iteration(next string) and similarly, second if-continue states that if similar string is found, we add this to `error_messages` and then continue with next string.

Reply:
It is very similar to the if-return from the `visit_call` method, in `visit_call`, we were parsing each string in a method call and if same string was found we had to return from the call and continue with the next string. But, here we're parsing each string in a loop and so if same string is found we do not execute next line, instead, we continue to the next iteration i.e. next string. And similarly in the second if-continue, if similar-string is found, we add the message to `error_messages` dict and continue with the next string. And if both the conditions fail, we're left with a distinct string and we add it to `_dict_of_strings`.